### DISCUSSION TITLE
How to update a collection (update nested entity, remove entity, add entity) ?

### DISCUSSION THREAD

**Discussion by dungkaka** [mikro-orm/mikro-orm#4810] [General]
I have a Resource entity that contains Action collection:
```Typescript
@Entity()
export class Resource {
    @PrimaryKey({ type: types.smallint })
    id: number;

    @Property({ unique: true })
    resource: string;

    @Property()
    resourceName: string;

    @OneToMany({ entity: () => Action, mappedBy: "resource", cascade: [Cascade.ALL], orphanRemoval: true })
    actions = new Collection<Action>(this);
}

// which Action class
@Entity()
export class Action {
    @PrimaryKey({ type: types.smallint })
    id: number;

    @ManyToOne(() => Resource)
    resource: Ref<Resource>;

    @Property()
    action: string;

    @Property({ nullable: true, serializer: forceUndefined })
    actionName?: string;
}
```
Assume that, I found a resource with full initialization of action:
```Typescript
const resource = await this.em.findOne(Resource, { id }, { populate: ["actions"] });
// resource = { ..., actions: [
//                        { id: 1, action: "create", actionName: "Create book" },
//                        { id: 2, action: "read", actionName: "Read book" }
//                     ]
//                }
```
Now I have updateResourceDTO and I want to update actions collection:
```Typescript
const updateResourceDTO = {...,  actions: [
                       { id: 1, action: "create", actionName: "Modify create book" },
                       { action: "other_action", actionName: "Other action for book" }
                     ]
                }}

// What I want is actions collection update to (result in database): 
// resource = { ..., actions: [
//                        { id: 1, action: "create", actionName: "Modify create book" },
//                        { id: 3, action: "other_action", actionName: "Other action for book" }
//                     ]
//                }
```

The expected behavior should be:
- Action with id = 1 is update actionName from "Create book" to "Modify create book"
- Action with id = 2 will delete because of disconnect from actions collection
- New action created with id = 3.

I try some solution for doing that:
- First solution:
```Typescript
resource.assign({
   ...updateResourceDTO,
    actions: updateResourceDTO.actions?.map((action) => this.em.create(Action, action))
})
this.em.flush()
```
- Second solution
```Typescript
resource.actions.set(updateResourceDTO.actions?.map((action) => this.em.create(Action, action)));
this.em.flush();
```
Yeah, both of them create new "other_action" (okey), delete action with id=2 (okey), <strong>BUT NOT UPDATE ACTION WITH id = 1 </strong>.
I have try some way to do that, I found some solution from issue or discussion, I know "updateNestedEntity" is add default on v5, but I still can not solve my problem.

NodeJS: 16
MikroORM: 5.8.1
Framework: NestJS
Environment: MacOS

**Comment by B4nan** [comment]
Have you tried `resource.assign(updateResourceDTO)`? You shouldn't need to use `em.create` explicitly here (and it is actually wrong to use that for managed entities that should trigger update, it only works for you because they were already loaded into the context - and it won't produce any updates as it hydrates the entity with your data, you don't want that).

**Reply by dungkaka** [reply]
If I using wrap(resource).assign(updateResourceDTO). It cause error: update "action" set "resource_id" = 30 where "id" = '[object Object]' - invalid input syntax for type smallint: "[object Object].

Query log: update "action" set "resource_id" = 30 where "id" = '[object Object]'. [took 4 ms]

This is weird track, and I don't know why this error happens :(((((

**Reply by B4nan** [reply]
Can you provide a complete repro for that?

**Reply by dungkaka** [reply]
Okay, I will try repro. Before that, I want ask if there are there some way can be done for solving my problem:
In actions collection:
- Action with id = 1 is update actionName from "Create book" to "Modify create book"
- Action with id = 2 will delete because of disconnect from actions collection
- New action created with id = 3.

I mean, it satify 3 condition: Delete entity that disconnect, add new entity if no declare id, and update entity if declare id.
Is it contain in mikroorm document ?

**Reply by B4nan** [reply]
As I said, you should be doing just `entity.assign(dto)`, it already handles the mapping of child items.

**Reply by dungkaka** [reply]
My DTO is not extends PlainObject.
Fortunately, I solved my problem by extends PlainObject from document. I don't know why but it works.
